### PR TITLE
code-gen(virtual_machine_management/VmHostAffinityPolicy): ansible v2

### LINKED ISSUE #993: [Feat] Develop ansible module for VM Host Affinity Policy under VMM namespace
# Add PC v4 VM Host Affinity Policy management modules

## Modules

- **`ntnx_vm_host_affinity_policy_v2`** — Create, update, delete VM Host Affinity Policies (name, description, list of VM references, list of host references)
- **`ntnx_vm_host_affinity_policies_info_v2`** — Fetch a specific policy by `ext_id` or list all policies

## Requirements

- [ ] VM Host Affinity Policies: create, update, delete (name, description, VM references, host references)
- [ ] VM Host Affinity Policies: idempotency check on update
- [ ] VM Host Affinity Policy info: fetch by `ext_id` or list all policies

## Todo

- [ ] Client SDK (`ntnx_vmm_py_client`), `get_spec()`
- [ ] Ansible spec & spec validator
- [ ] Sanity test
- [ ] Integration test
- [ ] Documentation
- [ ] Examples

### PR DESCRIPTION
**Pull request by abhinavbansal29**
## Summary

- Add `ntnx_vm_host_affinity_policy_v2` module for Create, Update, and Delete operations on VM-host affinity policies with idempotency and check_mode support
- Add `ntnx_vm_host_affinity_policies_info_v2` module for Get (by ext_id) and List (with pagination) operations
- Add helpers and API client functions for VmHostAffinityPoliciesApi
- Register new modules in `meta/runtime.yml` action groups
- Include examples and integration tests with full coverage (check_mode, CRUD, idempotency, negative scenarios)

## Test plan

- [x] Integration tests pass (Create, Update, Delete, Get, List, check_mode, idempotency, negative)
- [x] Examples playbook passes end-to-end
- [x] black, isort, flake8 pass
- [x] ansible-lint passes

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by abhinavbansal29** [comment]
## Example Run Logs
<details><summary>Click to expand</summary>

```
$(cat /tmp/example_logs.txt)
```

</details>

**Comment by abhinavbansal29** [comment]
## Test Run Logs
<details><summary>Click to expand</summary>

```
$(tail -500 /tmp/test_logs.txt)
```

</details>

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+            - Each entry specifies a category by its external ID.
+            - VMs with these categories will be hosted on hosts that match C(host_categories).
+        required: false
+        type: list
+        elements: dict
+        suboptions:
+            ext_id:
+                description:
+                    - The external ID of the category.
+                required: true
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:51]: This is not required as per docs.

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            - List of category references for hosts that VMs should have affinity with.
+            - Each entry specifies a category by its external ID.
+        required: false
+        type: list
+        elements: dict
+        suboptions:
+            ext_id:
+                description:
+                    - The external ID of the category.
+                required: true
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:64]: This is not required as per docs.

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    nutanix_password: "{{ password }}"
+    validate_certs: false
+    name: my_host_affinity_policy
+    description: Policy to keep critical VMs on specific hosts
+    vm_categories:
+      - ext_id: "2f54419e-596d-4b34-aa8f-1a1e944ee7d7"
+    host_categories:
+      - ext_id: "8811743f-f3ea-463c-539a-8d6a7f69b8f5"
+    state: present
+    wait: true
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py]: I think we can remove wait from all the examples

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+    ext_id: "605a0cf9-d04e-3be7-911b-1e6f193f6eb9"
+    state: absent
+    wait: true
+"""
+
+
+RETURN = r"""
+response:
+    description:
+        - The response from the VM host affinity policy operation.
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:146]: Please write proper description, like if Create  and wait true then this otherwise...

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+    sample: "98b9dc89-be08-3c56-b554-692b8b676fd2"
+ext_id:
+    description: The external ID of the VM host affinity policy.
+    type: str
+    returned: always
+    sample: "98b9dc89-be08-3c56-b554-692b8b676fd2"
+changed:
+    description: Indicates whether the resource was changed.
+    type: bool
+    returned: always
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:178]: Please add sample for all

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+            msg="Api Exception raised while updating VM host affinity policy",
+        )
+
+    task_ext_id = resp.data.ext_id
+    result["task_ext_id"] = task_ext_id
+    result["response"] = strip_internal_attributes(resp.data.to_dict())
+    if task_ext_id and module.params.get("wait"):
+        wait_for_completion(module, task_ext_id)
+
+    updated_policy = get_vm_host_affinity_policy(module, api_instance, ext_id)
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py]: We should fetch only if wait is True

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        result["response"] = strip_internal_attributes(resp.to_dict())
+    result["changed"] = True
+
+
+def run_module():
+    module = BaseModuleV4(
+        argument_spec=get_module_spec(),
+        supports_check_mode=True,
+        required_if=[
+            ("state", "present", ("name", "ext_id"), True),
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:411]: I think name is not required in create.

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+        ],
+    )
+    if SDK_IMP_ERROR:
+        module.fail_json(
+            msg=missing_required_lib("ntnx_vmm_py_client"), exception=SDK_IMP_ERROR
+        )
+
+    remove_param_with_none_value(module.params)
+    result = {
+        "changed": False,
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_v2.py:422]: Please add failed.

File: plugins/modules/ntnx_vm_host_affinity_policies_info_v2.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    }
+ext_id:
+    description: The external ID of the policy.
+    type: str
+    returned: when fetching a single policy
+    sample: "98b9dc89-be08-3c56-b554-692b8b676fd2"
+changed:
+    description: Indicates whether any changes were made (always false for info modules).
+    type: bool
+    returned: always
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policies_info_v2.py:112]: Please add sample for all

File: plugins/modules/ntnx_vm_host_affinity_policies_info_v2.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+def run_module():
+    module = BaseInfoModule(
+        argument_spec=get_module_spec(),
+        supports_check_mode=False,
+        mutually_exclusive=[
+            ("ext_id", "filter"),
+        ],
+    )
+    remove_param_with_none_value(module.params)
+    result = {"changed": False, "error": None, "response": None}
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policies_info_v2.py]: Please add failed, I think we can remove error.

File: plugins/modules/ntnx_vm_host_affinity_policy_re_enforce_v2.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    returned: always
+    sample: "98b9dc89-be08-3c56-b554-692b8b676fd2"
+changed:
+    description: Indicates whether the resource was changed.
+    type: bool
+    returned: always
+error:
+    description: The error message if an error occurred.
+    type: str
+    returned: when an error occurs
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_re_enforce_v2.py:126]: Please add sample

File: plugins/modules/ntnx_vm_host_affinity_policy_re_enforce_v2.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+def run_module():
+    module = BaseModuleV4(
+        argument_spec=get_module_spec(),
+        supports_check_mode=True,
+    )
+
+    remove_param_with_none_value(module.params)
+    result = {
+        "changed": False,
+        "error": None,
```
Comment (abhinavbansal29) [inline plugins/modules/ntnx_vm_host_affinity_policy_re_enforce_v2.py]: Please add failed and remove error

File: examples/vmm_v2/vm_host_affinity_policies_v2.yml [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    - name: Create VM host affinity policy
+      nutanix.ncp.ntnx_vm_host_affinity_policy_v2:
+        name: "{{ policy_name }}"
+        description: "Example VM host affinity policy"
+        vm_categories:
+          - ext_id: "{{ vm_category_ext_id }}"
+        host_categories:
+          - ext_id: "{{ host_category_ext_id }}"
+        state: present
+        wait: true
```
Comment (abhinavbansal29) [inline examples/vmm_v2/vm_host_affinity_policies_v2.yml]: Remove wait

File: examples/vmm_v2/vm_host_affinity_policies_v2.yml [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+      register: create_result
+
+    - name: Print create result
```
Comment (abhinavbansal29) [inline examples/vmm_v2/vm_host_affinity_policies_v2.yml]: Please remove print

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 15]
```diff
@@ -0,0 +1,1333 @@
...
+  nutanix.ncp.ntnx_categories_v2:
+    key: "{{ random_name }}_haf_key_{{ item }}"
+    value: "{{ random_name }}_haf_value_{{ item }}"
+    description: "ansible host-affinity category"
+  register: category_create_result
+  ignore_errors: true
+  loop: "{{ range(0, 5) | list }}"
+
+- name: Print categories creation result
+  ansible.builtin.debug:
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]: Remove all the debug statements.

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 16]
```diff
@@ -0,0 +1,1333 @@
...
+      failed"
+    success_msg:
+      "Verify generated spec for a VM host affinity policy using check
+      mode passed"
+
+##################################################################################
+# Create - Minimal spec
+##################################################################################
+
+- name: Create a VM host affinity policy with minimal spec
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml:359]: I think there is no point of minimal spec here, as we don't have any required field.

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 17]
```diff
@@ -0,0 +1,1333 @@
...
+  nutanix.ncp.ntnx_vm_host_affinity_policy_v2:
+    state: present
+    name: "{{ policy_name }}_1"
+    vm_categories:
+      - ext_id: "{{ categories_ext_ids[0] }}"
+    host_categories:
+      - ext_id: "{{ categories_ext_ids[0] }}"
+    wait: true
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]: Please remove wait

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 18]
```diff
@@ -0,0 +1,1333 @@
...
+    todelete: '{{ todelete + [result["response"]["ext_id"]] }}'
+
+##################################################################################
+# Create - Full spec
+##################################################################################
+
+- name: Create a VM host affinity policy with full spec
+  nutanix.ncp.ntnx_vm_host_affinity_policy_v2:
+    state: present
+    name: "{{ policy_name }}_3"
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]:
```suggestion
    name: "{{ policy_name }}_2"
```

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 19]
```diff
@@ -0,0 +1,1333 @@
...
+  ansible.builtin.debug:
+    var: result
+
+- name: Verify created VM host affinity policy with full spec
+  ansible.builtin.assert:
+    that:
+      - result.changed == true
+      - result.failed == false
+      - result.response is defined
+      - result.response.name == "{{ policy_name }}_3"
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]:
```suggestion
      - result.response.name == "{{ policy_name }}_2"
```

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 20]
```diff
@@ -0,0 +1,1333 @@
...
+
+##################################################################################
+# Create - Full spec
+##################################################################################
+
+- name: Create a VM host affinity policy with full spec
+  nutanix.ncp.ntnx_vm_host_affinity_policy_v2:
+    state: present
+    name: "{{ policy_name }}_3"
+    description: "{{ policy_name }}_3_description"
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]:
```suggestion
    description: "{{ policy_name }}_2_description"
```

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 21]
```diff
@@ -0,0 +1,1333 @@
...
+    var: result
+
+- name: Verify created VM host affinity policy with full spec
+  ansible.builtin.assert:
+    that:
+      - result.changed == true
+      - result.failed == false
+      - result.response is defined
+      - result.response.name == "{{ policy_name }}_3"
+      - result.response.description == "{{ policy_name }}_3_description"
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]:
```suggestion
      - result.response.description == "{{ policy_name }}_2_description"
```

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 22]
```diff
@@ -0,0 +1,1333 @@
...
+  ansible.builtin.set_fact:
+    todelete: '{{ todelete + [result["response"]["ext_id"]] }}'
+
+##################################################################################
+# Update - Check mode
+##################################################################################
+
+- name:
+    Generate spec for updating an existing VM host affinity policy using check
+    mode
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml]: Why are we writing in 3 lines?

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml [hunk 23]
```diff
@@ -0,0 +1,1333 @@
...
+      - result.response is defined
+      - result.response | length == 1
+    fail_msg: "Verify list of VM host affinity policies with limit 1 failed"
+    success_msg: "Verify list of VM host affinity policies with limit 1 passed"
+
+##################################################################################
+# Negative - Create without name
+##################################################################################
+
+- name: Create VM host affinity policy without name (negative test)
```
Comment (abhinavbansal29) [inline tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml:627]: Please check in the API docs, if name is required or not.

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py (hunk 1 above)
Comment (george-ghawali) [reply plugins/modules/ntnx_vm_host_affinity_policy_v2.py:51]: as per API docs
[ 1 .. 50 ]
at least one element should be provided

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py (hunk 7 above)
Comment (george-ghawali) [reply plugins/modules/ntnx_vm_host_affinity_policy_v2.py:411]: "message": "Failed to create the policy due to an invalid argument with key 'name' and value 'EMPTY'."

File: plugins/modules/ntnx_vm_host_affinity_policy_v2.py (hunk 2 above)
Comment (george-ghawali) [reply plugins/modules/ntnx_vm_host_affinity_policy_v2.py:64]: as per API docs
[ 1 .. 50 ]
at least one element should be provided

File: tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml (hunk 16 above)
Comment (george-ghawali) [reply tests/integration/targets/ntnx_vm_host_affinity_policy_v2/tasks/vm_host_affinity_policy_operations.yml:359]: name is required
